src/backend/distributed/executor/executor.c

```diff
@@ -779,6 +783,57 @@ LockPartitionsForDistributedPlan(DistributedPlan *distributedPlan)
 }
 
 
+/*
+ * AcquireExecutorShardLocks acquires advisory lock on shard IDs to prevent
+ * unsafe concurrent modifications of shards.
+ *
+ * We prevent concurrent modifications of shards in two cases:
+ * 1. Any non-commutative writes to a replicated table
+ * 2. Multi-shard writes that are executed in parallel
+ *
+ * The first case ensures we do not apply updates in different orders on
+ * different replicas (e.g. of a reference table), which could lead the
+ * replicas to diverge.
+ *
+ * The second case prevents deadlocks due to out-of-order execution.
+ *
+ * We do not take executor shard locks for utility commands such as
+ * TRUNCATE because the table locks already prevent concurrent access.
+ */
+static void
+AcquireExecutorShardLocks(DistributedExecution *execution)
+{
+	CmdType operation = execution->operation;
+	List *taskList = execution->tasksToExecute;
+
+	if (!(operation == CMD_INSERT || operation == CMD_UPDATE || operation == CMD_DELETE))
+	{
+		/* executor locks only apply to DML commands */
+		return;
+	}
+
+	/*
+	 * When executing in sequential mode or only executing a single task, we
+	 * do not need multi-shard locks.
+	 */
+	if (list_length(taskList) == 1 || ShouldRunTasksSequentially(taskList))
```

Comment:
Self note to myself (and please confirm that my understanding is correct about our conversations). Later, once I compile all the behavioral changes in a document, I'd like to add this as well.

We're slightly changing two behaviours which are also related to here:

- When ` SET citus.multi_shard_modify_mode TO sequential`, the new executor executes the multi shard UPDATE/DELETEs in parallel across worker nodes. On the other hand,  router executor serializes the execution across worker nodes as well. Router executor's behaviour is not intended for multi shard DMLs, mostly it was simpler to implement at that time and we haven't considered this use-case useful ever (for multi-row INSERTs it was intended and we're already enforcing it here). 
- With router executor, we're not acquiring the `AcquireExecutorMultiShardLocks ` for the above case. But, now we're acquiring the locks. The reason is actually above. Because with this new behaviour, the user could get into distributed deadlocks across concurrent multi-shard DMLs.

But, given that running multi-shard DMLs with sequential mode is mostly** relevant to supporting foreign keys from distributed tables, we actually do not care much about it right now. We can revisit this idea if anyone complains about it.

Finally, with fkey to reference table case, the new approach would increase the performance in many cases and doesn't lead to any potential distributed deadlock. One super edge case is that if the executor switches to sequential mode and then a distributed DML comes in the transaction block, it'd acquire the executor locks and the other concurrent parallel DMLs on the same table might need to wait for the locks. But, that seems very unlikely situation. 


** we've not heard any user trying to execute multi-shard DMLs in sequential mode. In fact, that's not a good idea.